README.md
# KP Player Menu
[![GitHub license](https://img.shields.io/github/license/KillahPotatoes/KP-PLM.svg)](https://github.com/KillahPotatoes/KP-PLM/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/KillahPotatoes/KP-PLM.svg)](https://github.com/KillahPotatoes/KP-PLM/releases)
[![GitHub Release Date](https://img.shields.io/github/release-date/KillahPotatoes/KP-PLM.svg)](https://github.com/KillahPotatoes/KP-PLM/releases)
[![Github All Releases](https://img.shields.io/github/downloads/KillahPotatoes/KP-PLM/total.svg)](https://github.com/KillahPotatoes/KP-PLM)

[![GitHub issues](https://img.shields.io/github/issues-raw/KillahPotatoes/KP-PLM.svg)](https://github.com/KillahPotatoes/KP-PLM/issues)
[![GitHub closed issues](https://img.shields.io/github/issues-closed-raw/KillahPotatoes/KP-PLM.svg)](https://github.com/KillahPotatoes/KP-PLM/issues?q=is%3Aissue+is%3Aclosed)
[![GitHub pull requests](https://img.shields.io/github/issues-pr-raw/KillahPotatoes/KP-PLM.svg)](https://github.com/KillahPotatoes/KP-PLM/pulls)
[![GitHub closed pull requests](https://img.shields.io/github/issues-pr-closed-raw/KillahPotatoes/KP-PLM.svg)](https://github.com/KillahPotatoes/KP-PLM/pulls?q=is%3Apr+is%3Aclosed)

[Discord](https://discord.gg/fjSPn8t)

If you like the work and think it's worth a small donation, feel free to use the following link:

[Donate via paypal.me](https://www.paypal.me/wyqer)

## Description
![plmpreview](https://user-images.githubusercontent.com/3811977/43684198-e4824120-989b-11e8-8261-26332fecd2e4.png)

With this dialog you can give the players of your mp mission the possibility to use their KP Player Menu settings, quality of life functionalities and the own lightweight group management. In the first place this dialog was created for the [KP Liberation Mission](https://github.com/KillahPotatoes/KP-Liberation), but we decided to provide it also as standalone version for other mission creators.

They can adjust their view distances depending if they're on foot, in a ground or air vehicle, as this is capped to 1600m in a normal mp mission, ignoring the players graphic settings. Same goes for the terrain detail, which is normally set to a value of 10 (that's basically between high and very high) in a mp mission by default. Also some new features like auto switch to 3rd person and disable radio chatter of AI units in your squad. As this player menu saves the settings in the player profile, the player will have the same settings on all servers/missions which use this menu. It also supports KP Ranks by providing a small overview about your progress and current state, but KP Ranks isn't needed at all to use the features.

## Features
* A lightweight group management
* Individual view distances
* Change of the terrain detail (like deactivating the grass)
* Automatic 3rd person view switch when entering a vehicle
* AI radio chatter disable (disable just the voice or the voice and chat)
* Adjustment of the in vehicle sound volume
* Persistent settings, which will apply for the player on each server and each mission who uses this menu

## Languages
Currently the plugin is localized for:
* Chinese Traditional
* English
* German
* Portuguese

## Mods
Required:
* [CBA_A3](https://steamcommunity.com/workshop/filedetails/?id=450814997)

## How to use
After downloading the latest release file from the [Releases Tab](https://github.com/KillahPotatoes/KP-PLM/releases) and extracting the rar archive you have to move the folders `KPGUI` and `KPPLM` to the root of your mission folder.

e.g.: `myFancyMission.Altis\KPGUI\` and `myFancyMission.Altis\KPPLM\`

The `stringtable.xml` goes also into the root of the mission folder.

If you don't already have a `description.ext` file you can just move the one from the release package to the root of your mission folder.

Should you already have an own `description.ext` file, add the following:

    #include "KPGUI\KPGUI_defines.hpp"
    #include "KPPLM\ui\KPPLM_dialog.hpp"

If you don't have a `CfgFunctions` definition in your `description.ext` you can append this:

    class CfgFunctions {
        #include "KPPLM\KPPLM_functions.hpp"
    };

If you have own defined functions in your `description.ext` you add at the line before the closing bracket just the `#include "KPPLM\KPPLM_functions.hpp"`

That's it. Your mission will now use and support the KP Player Menu.

## Licence
Copyright (C) 2018 [Wyqer](https://github.com/Wyqer)

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program. If not, see http://www.gnu.org/licenses/.


CHANGELOG.md
# Changelog

## 1.0.2
* Added: Portuguese localization. Thanks to [AngusDLX](https://github.com/AngusDLX)
* Added: Chinese Traditional localization. Thanks to [dustin902001](https://github.com/dustin902001)
* Added: Version icon in title bar.
* Updated: KPGUI defines.
* Tweaked: Initialization of the script.
* Tweaked: Use CBA function for persistent "open dialog" action on the player.

## 1.0.1
* Added: KP Ranks score value will be masked, if leveling system is disabled.
* Tweaked: Usage of KPR 2.0.2 functions.
* Fixed: Dialogs close directly if opened with "standard action" key.

## 1.0.0
* Initial release as standalone version. Outsourced from [KP Liberation](https://github.com/KillahPotatoes/KP-Liberation).


KPPLM/KPPLM_functions.hpp
/*
    KP PLAYER MENU FUNCTIONS

    File: KPPLM_functions.hpp
    Author: Wyqer - https://github.com/KillahPotatoes
    Date: 2018-08-06
    Last Update: 2018-08-07
    License: GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.html

    Description:
    Function defines for the KP Player Menu.
*/

class KPPLM {
    class main {
        file = "KPPLM\fnc";

        // Apply and save settings
        class apply {};

        // Creates a new group
        class create {};

        // Apply view distance and sound volume
        class getInOut {};

        // Initialize the KP player menu
        class init {
            postInit = 1;
        };

        // Joins selected group
        class join {};

        // Manages action adding and removing
        class manageActions {};

        // Open the dialog
        class openDialog {};

        // Applies new name to group
        class rename {};

        // Save dialog settings
        class save {};

        // Get and display KP Ranks data
        class showRankData {};
    };
};


KPPLM/ui/KPPLM_dialog.hpp
/*
    KP player menu dialog

    File: KPPLM_dialog.hpp
    Author: Wyqer - https://github.com/KillahPotatoes
    Date: 2018-08-03
    Last Update: 2018-11-10
    License: GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.html

    Description:
        Provides various functionalities for the players like group management, view distances, etc.
*/

class KPPLM_dialog {
    idd = 75803;
    movingEnable = 0;
    // Fetch all player groups when opened and free the variable when closed
    onLoad = "KPPLM_groups = allGroups select {isPlayer (leader _x)}";
    onUnload = "KPPLM_groups = nil";

    class controlsBackground {

        class KP_DialogTitle: KPGUI_PRE_DialogTitleC {
            text = "$STR_KPPLM_TITLE";
        };

        class KP_DialogArea: KPGUI_PRE_DialogBackgroundC {};

        class KP_LabelPlayer: KPGUI_PRE_InlineTitle {
            text = "$STR_KPPLM_LABELPLAYER";
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,0,16);
            w = KP_GETW(KP_WIDTH_VAL_C,1);
            h = KP_GETH(KP_HEIGHT_VAL_C,16);
        };

        class KP_TextRank: KPGUI_PRE_Text {
            idc = 758031;
            text = "$STR_KPPLM_RANK";
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,1,16);
            w = KP_GETW(KP_WIDTH_VAL_C,2);
            h = KP_GETH(KP_HEIGHT_VAL_C,24);
        };

        class KP_TextScore: KP_TextRank {
            idc = 758033;
            text = "$STR_KPPLM_SCORE";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,5,48);
        };

        class KP_TextPlaytime: KP_TextRank {
            idc = 758035;
            text = "$STR_KPPLM_PLAYTIME";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,7,48);
        };

        class KP_NoRanks: KP_LabelPlayer {
            idc = 758037;
            style = 2;
            colorText[] = {0.8, 0, 0, 0.75};
            text = "$STR_KPPLM_NORANKS";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,3,32);
            sizeEx = KP_TEXT_L;
        };

        class KP_LabelGroups: KP_LabelPlayer {
            text = "$STR_KPPLM_LABELGROUPS";
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,3,16);
            w = KP_GETW(KP_WIDTH_VAL_C,1);
            h = KP_GETH(KP_HEIGHT_VAL_C,16);
        };

        class KP_LabelView: KP_LabelPlayer {
            text = "$STR_KPPLM_LABELVIEW";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,1,2);
        };

        class KP_TextViewFoot: KPGUI_PRE_Text {
            text = "$STR_KPPLM_VIEWFOOT";
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,1,2);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,9,16);
            w = KP_GETW(KP_WIDTH_VAL_C,2);
            h = KP_GETH(KP_HEIGHT_VAL_C,24);
        };

        class KP_TextViewVehicle: KP_TextViewFoot {
            text = "$STR_KPPLM_VIEWVEHICLE";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,29,48);
        };

        class KP_TextViewAir: KP_TextViewFoot {
            text = "$STR_KPPLM_VIEWAIR";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,31,48);
        };

        class KP_LabelMisc: KP_LabelPlayer {
            text = "$STR_KPPLM_LABELMISC";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,11,16);
        };

        class KP_TextTerrain: KP_TextRank {
            text = "$STR_KPPLM_TEXTTERRAIN";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,12,16);
        };

        class KP_TextTPV: KP_TextRank {
            text = "$STR_KPPLM_TEXTTPV";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,19,24);
        };

        class KP_TextRadio: KP_TextRank {
            text = "$STR_KPPLM_TEXTRADIO";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,10,12);
        };

        class KP_TextSound: KP_TextRank {
            text = "$STR_KPPLM_TEXTSOUND";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,22,24);
        };
    };

    class controls {

        class KP_PlayerRank: KPGUI_PRE_Text {
            idc = 758032;
            style = 2;
            text = "";
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,1,2);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,1,16);
            w = KP_GETW(KP_WIDTH_VAL_C,2);
            h = KP_GETH(KP_HEIGHT_VAL_C,24);
        };

        class KP_PlayerScore: KP_PlayerRank {
            idc = 758034;
            text = "";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,5,48);
        };

        class KP_PlayerTime: KP_PlayerRank {
            idc = 758036;
            text = "";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,7,48);
        };

        class KP_GroupList: KPGUI_PRE_ListBox {
            idc = 758038;
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,4,16);
            w = KP_GETW(KP_WIDTH_VAL_C,(4/3));
            h = KP_GETH(KP_HEIGHT_VAL_C,(24/5));
            onLBSelChanged = "ctrlSetText [758039, groupId (KPPLM_groups select (lbCurSel 758038))]";
        };

        class KP_ButtonCreate: KPGUI_PRE_InlineButton {
            text = "$STR_KPPLM_CREATE";
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,3,4);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,4,16);
            w = KP_GETW(KP_WIDTH_VAL_C,4);
            h = KP_GETH(KP_HEIGHT_VAL_C,(48/5));
            onButtonClick = "[] call KPPLM_fnc_create";
        };

        class KP_ButtonJoin: KP_ButtonCreate {
            text = "$STR_KPPLM_JOIN";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,17,48);
            onButtonClick = "[] call KPPLM_fnc_join";
        };

        class KP_RenameEdit: KPGUI_PRE_EditBox {
            idc = 758039;
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,11,24);
            w = KP_GETW(KP_WIDTH_VAL_C,(4/3));
            h = KP_GETH(KP_HEIGHT_VAL_C,24);
        }

        class KP_ButtonRename: KP_ButtonCreate {
            text = "$STR_KPPLM_RENAME";
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,11,24);
            h = KP_GETH(KP_HEIGHT_VAL_C,24);
            onButtonClick = "[] call KPPLM_fnc_rename";
        };

        class KP_BoxViewFoot: KPGUI_PRE_EditBox {
            idc = 7580310;
            style = 65;
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,1,4);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,9,16);
            w = KP_GETW(KP_WIDTH_VAL_C,4);
            h = KP_GETH(KP_HEIGHT_VAL_C,24);
        };

        class KP_BoxViewVehicle: KP_BoxViewFoot {
            idc = 7580311;
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,29,48);
        };

        class KP_BoxViewAir: KP_BoxViewFoot {
            idc = 7580312;
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,31,48);
        };

        class KP_ComboTerrain: KPGUI_PRE_Combo {
            idc = 7580313;
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,1,2);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,12,16);
            w = KP_GETW(KP_WIDTH_VAL_C,2);
            h = KP_GETH(KP_HEIGHT_VAL_C,24);
        };

        class KP_ComboTPV: KP_ComboTerrain {
            idc = 7580314;
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,19,24);
        };

        class KP_ComboRadio: KP_ComboTerrain {
            idc = 7580315;
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,10,12);
        };

        class KP_TextSoundValue: KP_PlayerRank {
            idc = 7580316;
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,22,24);
        };

        class KP_SliderSound: KPGUI_PRE_Slider {
            idc = 7580317;
            x = KP_GETCX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
            y = KP_GETCY(KP_Y_VAL_C,KP_HEIGHT_VAL_C,23,24);
            w = KP_GETW(KP_WIDTH_VAL_C,1);
            h = KP_GETH(KP_HEIGHT_VAL_C,24);
            tooltip = "$STR_KPPLM_TTSOUND";
            onSliderPosChanged = "ctrlSetText [7580316, format ['%1%2', round (_this select 1), '%']]";
        };

        class KP_ApplyButton: KPGUI_PRE_DialogButtonC {
            text = "$STR_KPPLM_APPLY";
            onButtonClick = "[] call KPPLM_fnc_save";
        };

        class KP_DialogCross: KPGUI_PRE_DialogCrossC {};
    };
};


KPGUI/KP_uiDefines.hpp
/*
    Killah Potatoes GUI System

    File: KP_uiDefines.hpp
    Author: Wyqer - https://github.com/KillahPotatoes
    Date: 2018-07-10
    Last Update: 2018-08-07
    License: GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.html

    Description:
    UI defines and classes used in mods and missions from the Killah Potatoes Gaming Community.
*/

// BI class includes
// Details via: "Default" call BIS_fnc_exportGUIBaseClasses;
#include "A3_defines.hpp"

/*
    --- General Defines ---
*/

// Text sizes
#define KP_TEXT_XS                      safeZoneH * 0.016
#define KP_TEXT_S                       safeZoneH * 0.018
#define KP_TEXT_M                       safeZoneH * 0.02
#define KP_TEXT_L                       safeZoneH * 0.025
#define KP_TEXT_XL                      safeZoneH * 0.03

// Constant values for calculation
#define KP_SPACING_X                    0.002
#define KP_SPACING_Y                    0.004
#define KP_HEIGTH_TITLE                 0.035
#define KP_HEIGTH_BUTTON                0.025

// Colors
#define KP_COLOR_PLAYERDEFINE           {"(profilenamespace getvariable ['GUI_BCG_RGB_R',0.13])", "(profilenamespace getvariable ['GUI_BCG_RGB_G',0.54])", "(profilenamespace getvariable ['GUI_BCG_RGB_B',0.21])", "(profilenamespace getvariable ['GUI_BCG_RGB_A',0.8])"}
#define KP_COLOR_BACKGROUND             {0, 0, 0, 0.5}

/*
    --- Functions ---
*/

/*
    Get X coordinate
    Example: KP_GETCX(KP_X_VAL,KP_WIDTH_VAL,2,4)
    KP_X_VAL -> Left offset of the left dialog side from the edge of the screen
    KP_WIDTH_VAL -> Width of the whole dialog (KP_X_VAL spacing to the left and to the right)
    2 -> Coordinate of the 3rd element position (0,1,2,3,etc.)
    4 -> In relation to a 4 element X grid

    GETX is for a coordinate in relation to the whole dialog width
    GETCX is for a coordinate in relation to the content area (which has an equal spacing to all sides)
*/

#define KP_GETX(XVAL,WVAL,POS,GRID)     safeZoneX + safeZoneW * (XVAL + POS * (WVAL + KP_SPACING_X) / GRID)
#define KP_GETCX(XVAL,WVAL,POS,GRID)    safeZoneX + safeZoneW * (XVAL + (POS * WVAL + (GRID - POS) * KP_SPACING_X) / GRID)
#define KP_GETX_CROSS(XVAL)             safeZoneX + safeZoneW * (1 - XVAL - 0.02)

/*
    Get Y coordinate
    Example: KP_GETCY(KP_Y_VAL,KP_HEIGHT_VAL,3,8)
    KP_Y_VAL -> Top offset of the dialog top from the edge of the screen
    KP_HEIGHT_VAL -> Height of the area between title bar and bottom buttons in the dialog
    3 -> Coordinate of the 4th element position (0,1,2,3,etc.)
    8 -> In relation to a 8 element Y grid

    GETY_AREA is for the Y coordinate of the beginning of the background (which frames the content area)
    GETY_BELOW is for the Y coordinate of the buttons below the background
*/
#define KP_GETCY(YVAL,HVAL,POS,GRID)    safeZoneY + safeZoneH * ((YVAL + KP_HEIGTH_TITLE + KP_SPACING_Y) + (POS * HVAL + (GRID - POS) * KP_SPACING_Y) / GRID)
#define KP_GETY_CROSS(YVAL)             safeZoneY + safeZoneH * (YVAL + 0.005)
#define KP_GETY_AREA(YVAL)              safeZoneY + safeZoneH * (YVAL + KP_HEIGTH_TITLE + KP_SPACING_Y)
#define KP_GETY_BELOW(YVAL,HVAL)        safeZoneY + safeZoneH * (YVAL + KP_HEIGTH_TITLE + 2 * KP_SPACING_Y + HVAL)

/*
    Get width for an element
    Example: KP_GETW(KP_WIDTH_VAL,2)
    KP_WIDTH_VAL -> Width of the whole dialog (KP_X_VAL spacing to the left and to the right)
    2 -> Width for an element which fits 2 times in the content area from left to right (substracted by spacings to other elements left and right)

    GETW is in relation to the content area
    GETWPLAIN is in relation to the whole dialog width (for the buttons below for example)
*/
#define KP_GETW(WVAL,GRID)             safeZoneW * ((WVAL - (GRID + 1) * KP_SPACING_X) / GRID)
#define KP_GETWPLAIN(WVAL,GRID)        safeZoneW * ((WVAL - (GRID - 1) * KP_SPACING_X) / GRID)

/*
    Get height for an element
    Example: KP_GETH(KP_HEIGHT_VAL,10)
    KP_HEIGHT_VAL -> Height of the area between title bar and bottom buttons in the dialog
    10 -> Height for an element which fits 10 times in the content area from top to bottom (substracted by spacings to other elements above and below)
*/
#define KP_GETH(HVAL,GRID)             safeZoneH * ((HVAL - (GRID + 1) * KP_SPACING_Y) / GRID)

/*
    --- General Classes ---
*/

// Title bar for dialogs
class KP_Title: RscText {
    colorBackground[] = KP_COLOR_PLAYERDEFINE;
    text = "DIALOG TITLE";
    h = safeZoneH * KP_HEIGTH_TITLE;
    shadow = 1;
    sizeEx = KP_TEXT_XL;
};

// Cross symbol for dialog close
class KP_CloseCross: RscActiveText {
    style = 48;
    color[] = {1, 1, 1, 0.75};
    colorBackground[] = {0, 0, 0, 0};
    colorText[] = {1, 1, 1, 0.75};
    colorActive[] = {1, 1, 1, 1};
    text = "\A3\Ui_f\data\GUI\Rsc\RscDisplayArcadeMap\icon_exit_cross_ca.paa";
    w = safeZoneW * 0.014;
    h = safeZoneH * 0.025;
    tooltip = "$STR_DISP_CLOSE";
    action = "closeDialog 0";
};

// General background for the dialog area
class KP_Background: RscText {
    colorBackground[] = KP_COLOR_BACKGROUND;
};

// Normal Text
class KP_Text: RscText {
    sizeEx = KP_TEXT_M;
};

// Inline Title
class KP_InlineTitle: RscText {
    style = 2;
    sizeEx = KP_TEXT_L;
};

// Active Text
class KP_ActiveText: RscActiveText {
    style = 2;
    sizeEx = KP_TEXT_M;
};

// Picture
class KP_Picture: RscText {
    style = 48;
};

class KP_ActivePicture: RscActivePicture {};

// Picture which keeps aspect ratio
class KP_PictureRatio: RscText {
    style = 48 + 2048;
};

// Button
class KP_Button: RscButton {
    colorBackground[] = KP_COLOR_BACKGROUND;
    colorFocused[] = KP_COLOR_BACKGROUND;
    h = safeZoneH * KP_HEIGTH_BUTTON;
    sizeEx = KP_TEXT_M;
};

// Inline Button
class KP_InlineButton: RscButton {
    colorBackground[] = {0.25, 0.25, 0.25, 1};
    colorBackgroundDisabled[] = {0, 1, 0, 1};
    colorBackgroundActive[] = {0.3, 0.3, 0.3, 1};
    colorFocused[] = {0.25, 0.25, 0.25, 1};
    sizeEx = KP_TEXT_M;
    offsetPressedX = safeZoneW * 0.0005;
    offsetPressedY = safeZoneH * 0.001;
}

// Controls group
class KP_ControlsGroup: RscControlsGroup {};

// Combo
class KP_Combo: RscCombo {
    colorBackground[] = KP_COLOR_BACKGROUND;
    sizeEx = KP_TEXT_M;
};

// Checkbox
class KP_CheckBox: RscCheckBox {
    style = 2;
    w = safeZoneW * 0.014;
    h = safeZoneH * 0.025;
};

// ListBox
class KP_ListBox: RscListBox {
    sizeEx = KP_TEXT_M;
};

class KP_ListNBox: RscListNBox {
    sizeEx = KP_TEXT_M;
};

// XListBox
class KP_XListBox: RscXListBox {
    colorBackground[] = KP_COLOR_BACKGROUND;
    sizeEx = KP_TEXT_M;
};

// Slider
class KP_Slider: RscXSliderH {
    color[] = {1, 1, 1, 0.6};
    colorActive[] = {1, 1, 1, 0.6};
};

// Editbox
class KP_EditBox: RscEdit {
    colorBackground[] = {0, 0, 0, 0.3};
    sizeEx = KP_TEXT_M;
};

/*
    --- Standard sized dialog components ---
    (X from 0.25 - 0.75, Y from 0.2 - 0.8)
*/

#define KP_X_VAL                        0.25
#define KP_Y_VAL                        0.2

#define KP_WIDTH_VAL                    (1 - 2 * KP_X_VAL)
#define KP_HEIGHT_VAL                   (1 - 2 * KP_Y_VAL - KP_HEIGTH_TITLE - KP_HEIGTH_BUTTON - 2 * KP_SPACING_Y)

// Title bar
class KP_DialogTitle: KP_Title {
    x = KP_GETX(KP_X_VAL,KP_WIDTH_VAL,0,1);
    y = safeZoneY + safeZoneH * KP_Y_VAL;
    w = KP_GETWPLAIN(KP_WIDTH_VAL,1);
};

// Cross symbol
class KP_DialogCross: KP_CloseCross {
    x = KP_GETX_CROSS(KP_X_VAL);
    y = KP_GETY_CROSS(KP_Y_VAL);
};

// Background
class KP_DialogBackground: KP_Background {
    x = KP_GETX(KP_X_VAL,KP_WIDTH_VAL,0,1);
    y = KP_GETY_AREA(KP_Y_VAL);
    w = KP_GETWPLAIN(KP_WIDTH_VAL,1);
    h = safeZoneH * KP_HEIGHT_VAL;
};

// Button pos 1 in 4 button row
class KP_DialogButton: KP_Button {
    x = KP_GETX(KP_X_VAL,KP_WIDTH_VAL,0,4);
    y = KP_GETY_BELOW(KP_Y_VAL,KP_HEIGHT_VAL);
    w = KP_GETWPLAIN(KP_WIDTH_VAL,4);
};

/*
    --- Large sized dialog components ---
    (X from 0.15 - 0.85, Y from 0.15 - 0.85)
*/

#define KP_X_VAL_L                      0.15
#define KP_Y_VAL_L                      0.15

#define KP_WIDTH_VAL_L                  (1 - 2 * KP_X_VAL_L)
#define KP_HEIGHT_VAL_L                 (1 - 2 * KP_Y_VAL_L - KP_HEIGTH_TITLE - KP_HEIGTH_BUTTON - 2 * KP_SPACING_Y)

// Title bar
class KP_DialogTitleL: KP_Title {
    x = KP_GETX(KP_X_VAL_L,KP_WIDTH_VAL_L,0,1);
    y = safeZoneY + safeZoneH * KP_Y_VAL_L;
    w = KP_GETWPLAIN(KP_WIDTH_VAL_L,1);
};

// Cross symbol
class KP_DialogCrossL: KP_CloseCross {
    x = KP_GETX_CROSS(KP_X_VAL_L);
    y = KP_GETY_CROSS(KP_Y_VAL_L);
};

// Background
class KP_DialogBackgroundL: KP_Background {
    x = KP_GETX(KP_X_VAL_L,KP_WIDTH_VAL_L,0,1);
    y = KP_GETY_AREA(KP_Y_VAL_L);
    w = KP_GETWPLAIN(KP_WIDTH_VAL_L,1);
    h = safeZoneH * KP_HEIGHT_VAL_L;
};

// Button pos 1 in 4 button row
class KP_DialogButtonL: KP_Button {
    x = KP_GETX(KP_X_VAL_L,KP_WIDTH_VAL_L,0,4);
    y = KP_GETY_BELOW(KP_Y_VAL_L,KP_HEIGHT_VAL_L);
    w = KP_GETWPLAIN(KP_WIDTH_VAL_L,4);
};

/*
    --- Corner dialog components ---
    (X from 0.035 - 0.235, Y from 0.05 - 0.8)
*/

#define KP_X_VAL_C                      0.035
#define KP_Y_VAL_C                      0.05

#define KP_WIDTH_VAL_C                  0.2
#define KP_HEIGHT_VAL_C                 (0.75 - KP_HEIGTH_TITLE - KP_HEIGTH_BUTTON - 2 * KP_SPACING_Y)

// Title bar
class KP_DialogTitleC: KP_Title {
    x = KP_GETX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
    y = safeZoneY + safeZoneH * KP_Y_VAL_C;
    w = KP_GETWPLAIN(KP_WIDTH_VAL_C,1);
};

// Cross symbol
class KP_DialogCrossC: KP_CloseCross {
    x = safeZoneX + safeZoneW * (KP_X_VAL_C + KP_WIDTH_VAL_C - 0.02)
    y = KP_GETY_CROSS(KP_Y_VAL_C);
};

// Background
class KP_DialogBackgroundC: KP_Background {
    x = KP_GETX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
    y = KP_GETY_AREA(KP_Y_VAL_C);
    w = KP_GETWPLAIN(KP_WIDTH_VAL_C,1);
    h = safeZoneH * KP_HEIGHT_VAL_C;
};

// Button
class KP_DialogButtonC: KP_Button {
    x = KP_GETX(KP_X_VAL_C,KP_WIDTH_VAL_C,0,1);
    y = KP_GETY_BELOW(KP_Y_VAL_C,KP_HEIGHT_VAL_C);
    w = KP_GETWPLAIN(KP_WIDTH_VAL_C,1);
};
